File: backend/routes/judges.js
import { Router } from 'express'
import { getDb } from '../db.js'

const router = Router()

// GET /judges
router.get('/judges', (req, res) => {
  const db = getDb()
  const rows = db.prepare('SELECT * FROM judges ORDER BY id DESC').all()
  for (const row of rows) {
    row.specializations = JSON.parse(row.specializations || '[]')
    row.available_days  = JSON.parse(row.available_days  || '[]')
  }
  res.json({ success: true, data: rows })
})

// GET /judges/:id
router.get('/judges/:id', (req, res) => {
  const db = getDb()
  const row = db.prepare('SELECT * FROM judges WHERE id = ?').get(Number(req.params.id))
  if (!row) return res.status(404).json({ success: false, error: { code: 'NOT_FOUND', message: 'Judge not found.' } })
  row.specializations = JSON.parse(row.specializations || '[]')
  row.available_days  = JSON.parse(row.available_days  || '[]')
  res.json({ success: true, data: row })
})

// POST /judges
router.post('/judges', (req, res) => {
  const db = getDb()
  const { name, email, phone, rank, gender, division, specializations, available_days, max_hearings_per_day } = req.body
  if (!name) return res.status(400).json({ success: false, error: { code: 'VALIDATION_ERROR', message: 'Name is required.' } })

  const info = db.prepare(
    `INSERT INTO judges (name, email, phone, rank, gender, division, specializations, available_days, max_hearings_per_day)
     VALUES (?,?,?,?,?,?,?,?,?)`
  ).run(
    name, email || '', phone || '', rank || 'High Court Judge', gender || 'Male',
    division || 'Abuja',
    JSON.stringify(specializations || []), JSON.stringify(available_days || []),
    max_hearings_per_day || 4
  )

  db.prepare("INSERT INTO audit_logs (user, action, target, time) VALUES (?,?,?,datetime('now'))")
    .run(req.user?.username || 'system', 'CREATE_JUDGE', `judge:${name}`)

  const row = db.prepare('SELECT * FROM judges WHERE id = ?').get(info.lastInsertRowid)
  row.specializations = JSON.parse(row.specializations || '[]')
  row.available_days  = JSON.parse(row.available_days  || '[]')
  res.status(201).json({ success: true, data: row })
})

// PUT /judges/:id
router.put('/judges/:id', (req, res) => {
  const db = getDb()
  const id = Number(req.params.id)
  const existing = db.prepare('SELECT * FROM judges WHERE id = ?').get(id)
  if (!existing) return res.status(404).json({ success: false, error: { code: 'NOT_FOUND', message: 'Judge not found.' } })

  const { name, email, phone, rank, gender, division, specializations, available_days, max_hearings_per_day, is_active } = req.body
  db.prepare(
    `UPDATE judges SET
       name=COALESCE(?, name), email=COALESCE(?, email), phone=COALESCE(?, phone),
       rank=COALESCE(?, rank), gender=COALESCE(?, gender), division=COALESCE(?, division),
       specializations=COALESCE(?, specializations), available_days=COALESCE(?, available_days),
       max_hearings_per_day=COALESCE(?, max_hearings_per_day), is_active=COALESCE(?, is_active)
     WHERE id = ?`
  ).run(
    name, email, phone, rank, gender, division,
    specializations ? JSON.stringify(specializations) : null,
    available_days  ? JSON.stringify(available_days)   : null,
    max_hearings_per_day, is_active, id
  )

  db.prepare("INSERT INTO audit_logs (user, action, target, time) VALUES (?,?,?,datetime('now'))")
    .run(req.user?.username || 'system', 'UPDATE_JUDGE', `judge:${existing.name}`)

  const row = db.prepare('SELECT * FROM judges WHERE id = ?').get(id)
  row.specializations = JSON.parse(row.specializations || '[]')
  row.available_days  = JSON.parse(row.available_days  || '[]')
  res.json({ success: true, data: row })
})

// DELETE /judges/:id
router.delete('/judges/:id', (req, res) => {
  const db = getDb()
  const id  = Number(req.params.id)
  const row = db.prepare('SELECT * FROM judges WHERE id = ?').get(id)
  if (!row) return res.status(404).json({ success: false, error: { code: 'NOT_FOUND', message: 'Judge not found.' } })

  db.prepare('DELETE FROM judges WHERE id = ?').run(id)
  db.prepare("INSERT INTO audit_logs (user, action, target, time) VALUES (?,?,?,datetime('now'))")
    .run(req.user?.username || 'system', 'DELETE_JUDGE', `judge:${row.name}`)

  res.json({ success: true, data: { id } })
})

export default router

File: backend/api/judges.py
"""
api/judges.py
─────────────
Judge management endpoints — response fields mapped to match frontend expectations.
"""

from fastapi         import APIRouter, Depends, HTTPException
from sqlalchemy.orm  import Session
from typing          import List, Optional

from db.database     import get_db
from db.crud         import (get_all_judges, get_judge_by_id, get_judge_by_code,
                              create_judge, update_judge, delete_judge)
from core.security   import get_current_user

router = APIRouter(prefix="/api/judges", tags=["Judges"])


def _format_judge(judge):
    """Transform DB judge object to match frontend field expectations."""
    specs = [s.strip() for s in judge.specialisations.split(",") if s.strip()]
    return {
        "id":                   judge.id,
        "judge_code":           judge.judge_code,
        "name":                 judge.name,
        "email":                f"{judge.judge_code.lower()}@court.gov.ng",
        "phone":                "",
        "rank":                 "High Court Judge",
        "gender":               "Male",
        "division":             judge.court_location or "Abuja",
        "specializations":      specs,          # array, not comma string
        "specialisations":      judge.specialisations,  # keep original too
        "available_days":       ["monday","tuesday","wednesday","thursday","friday"],
        "max_hearings_per_day": 4,
        "is_available":         judge.is_available,
        "is_active":            judge.is_available,
        "court_location":       judge.court_location,
        "created_at":           judge.created_at.isoformat(),
    }


@router.get("", response_model=None, summary="List all judges")
@router.get("/", response_model=None, summary="List all judges", include_in_schema=False)
def list_judges(db: Session = Depends(get_db)):
    return [_format_judge(j) for j in get_all_judges(db)]


@router.get("/{judge_id}", response_model=None, summary="Get a judge by ID")
def get_judge(judge_id: str, db: Session = Depends(get_db)):
    judge = get_judge_by_id(db, judge_id)
    if not judge:
        raise HTTPException(status_code=404, detail="Judge not found.")
    return _format_judge(judge)


@router.post("", response_model=None, status_code=201, summary="Add a new judge")
@router.post("/", response_model=None, status_code=201, include_in_schema=False)
async def add_judge(request_data: dict,
                    db: Session = Depends(get_db),
                    current_user=Depends(get_current_user)):
    # Accept both 'specializations' (array) and 'specialisations' (string)
    specs = request_data.get("specializations") or request_data.get("specialisations", [])
    if isinstance(specs, str):
        specs = [s.strip() for s in specs.split(",")]

    # Generate judge code
    existing = get_all_judges(db)
    code = f"J{str(len(existing)+1).zfill(3)}"

    if get_judge_by_code(db, code):
        code = f"J{str(len(existing)+100).zfill(3)}"

    judge = create_judge(
        db,
        judge_code     = code,
        name           = request_data.get("name",""),
        specialisations= specs,
        court_location = request_data.get("division") or request_data.get("court_location",""),
    )
    return _format_judge(judge)


@router.put("/{judge_id}", response_model=None, summary="Update judge")
@router.patch("/{judge_id}", response_model=None, summary="Update judge", include_in_schema=False)
async def edit_judge(judge_id: str,
                     request_data: dict,
                     db: Session = Depends(get_db),
                     current_user=Depends(get_current_user)):
    specs = request_data.get("specializations") or request_data.get("specialisations")
    if specs and isinstance(specs, list):
        request_data["specialisations"] = specs

    location = request_data.get("division") or request_data.get("court_location")
    if location:
        request_data["court_location"] = location

    # Map is_active → is_available
    if "is_active" in request_data:
        request_data["is_available"] = request_data.pop("is_active")

    judge = update_judge(db, judge_id, **{
        k: v for k, v in request_data.items()
        if k in ["name","specialisations","is_available","court_location"]
    })
    if not judge:
        raise HTTPException(status_code=404, detail="Judge not found.")
    return _format_judge(judge)


@router.delete("/{judge_id}", summary="Remove a judge")
def remove_judge(judge_id: str,
                 db: Session = Depends(get_db),
                 current_user=Depends(get_current_user)):
    if not delete_judge(db, judge_id):
        raise HTTPException(status_code=404, detail="Judge not found.")
    return {"message": "Judge deleted successfully."}